src/core/lib/security/security_connector/tls/spiffe_security_connector.cc

```diff
@@ -0,0 +1,465 @@
+        key_materials_config.get()->pem_key_cert_pair_list().size();
+    grpc_security_status status = grpc_init_tsi_ssl_server_handshaker_factory(
+        pem_key_cert_pairs, num_key_cert_pairs,
+        key_materials_config.get()->pem_root_certs(),
+        creds->options()->cert_request_type(), &server_handshaker_factory_);
+    // Free memory.
+    key_materials_config.get()->Unref();
+    grpc_tsi_ssl_pem_key_cert_pairs_destroy(pem_key_cert_pairs,
+                                            num_key_cert_pairs);
+    return status;
+  }
+
+  void add_handshakers(grpc_pollset_set* interested_parties,
+                       grpc_core::HandshakeManager* handshake_mgr) override {
+    /* Create a TLS SPIFFE TSI handshaker for server. */
+    RefreshServerHandshakerFactory();
+    tsi_handshaker* tsi_hs = nullptr;
+    tsi_result result = tsi_ssl_server_handshaker_factory_create_handshaker(
+        server_handshaker_factory_, &tsi_hs);
+    if (result != TSI_OK) {
+      gpr_log(GPR_ERROR, "Handshaker creation failed with error %s.",
+              tsi_result_to_string(result));
+      return;
+    }
+    handshake_mgr->Add(grpc_core::SecurityHandshakerCreate(tsi_hs, this));
+  }
+
+  void check_peer(tsi_peer peer, grpc_endpoint* ep,
+                  grpc_core::RefCountedPtr<grpc_auth_context>* auth_context,
+                  grpc_closure* on_peer_checked) override {
+    grpc_error* error = grpc_ssl_check_peer(nullptr, &peer, auth_context);
+    tsi_peer_destruct(&peer);
+    GRPC_CLOSURE_SCHED(on_peer_checked, error);
+  }
+
+  int cmp(const grpc_security_connector* other) const override {
+    return server_security_connector_cmp(
+        static_cast<const grpc_server_security_connector*>(other));
+  }
+
+ private:
+  tsi_ssl_server_handshaker_factory* server_handshaker_factory_;
+};
+
+/* gRPC-provided callback executed by application, which servers to bring the
+ * control back to gRPC core. */
+static void ServerAuthorizationCheckDone(
+    grpc_tls_server_authorization_check_arg* arg) {
+  GPR_ASSERT(arg != nullptr);
+  grpc_core::ExecCtx exec_ctx;
+  grpc_error* error = ProcessServerAuthorizationCheckResult(arg);
+  grpc_tls_spiffe_channel_security_connector* sc =
+      static_cast<grpc_tls_spiffe_channel_security_connector*>(
+          arg->cb_user_data);
+  GRPC_CLOSURE_SCHED(const_cast<grpc_closure*>(sc->on_peer_checked()), error);
+}
+
+}  // namespace
+
+/** Create a gRPC TLS SPIFFE channel security connector. */
+grpc_core::RefCountedPtr<grpc_channel_security_connector>
+grpc_tls_spiffe_channel_security_connector_create(
+    grpc_core::RefCountedPtr<grpc_channel_credentials> channel_creds,
+    grpc_core::RefCountedPtr<grpc_call_credentials> request_metadata_creds,
+    const char* target_name, const char* overridden_target_name,
+    tsi_ssl_session_cache* ssl_session_cache) {
+  if (channel_creds == nullptr || target_name == nullptr) {
+    gpr_log(GPR_ERROR,
+            "Invalid arguments to "
```

Comment:
nit: this error is too vague, be specific

Reply:
Done.